### PR TITLE
[r] Add SOMA Experiment Query -> SingleCellExperiment outgestor

### LINKED ISSUE #1276: [r] Support Bioconductor query/outgest

### PR DESCRIPTION
**Pull request by mojaveazure**
Add support for creating `SingleCellExperiment` objects from SOMA experiment queries; all done using SingleCellExperiment as a suggested depdendency

Implemented SOMA methods:
- `SOMAExperimentAxisQuery$to_single_cell_experiment()`: allows loading in entire `SingleCellExperiment` objects

Future work:
- Add `SingleCellExperiment` -> SOMA ingestor via `write_soma()`
- Port `$to_single_cell_experiment()` to the `SOMAExperiment` level to allow loading of alternate experiments

Notes:
~Will not merge until support for core 2.16 is merged in~ Resolved

resolves #1276

### CODE REVIEW

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:731]: Is there _never_ `varm`??

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:767]: I know you know what this means

Please add a sentence or two for the benefit of the rest of us :)

File: apis/r/R/utils-sce.R
Comment (johnkerl) [inline apis/r/R/utils-sce.R:1]: Let's please rename to `utils-single-cell-experiment.R`

**Review by johnkerl — COMMENTED** [review]
LGTM (modulo comments below), although I'm not R-tastic enough to accept. (I will say unit-test coverage looks quite good :) )

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:772]: ❤️

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [reply apis/r/R/SOMAExperimentAxisQuery.R:731]: OK let's at least put a code comment saying it can't exist, and its omission is by thoughtful design.

We live in a polyglot world, and this is a polyglot project, and anyone with less R/Seurat expertise may be unnecessarily confused/concerned.

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:807]: Please spell out whatever `rd` is standing for

You do save yourself a few keystrokes typing it out here, as a one-time cost, but the added clarity to every reader forever after will be a net time savings

Comment (johnkerl) [reply apis/r/R/SOMAExperimentAxisQuery.R:807]: ... looks like it's `skip_reduced_dims`

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:824]: I know my feedback is delayed and I should have asked sooner but this is a wonderful opportunity for some simple little helper methods -- it would be easier to review, for sure, but also easier on future maintainers

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:864]: `skip_col_pairs` please

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:896]: `skip_row_pairs` please

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:933]: as above: let's make more of these convenience methods

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:1061]: `df` -> `df_name` -- `df` makes it sound like it _is_ a dataframe (it isn't) leading to a little unnecessary cognitive hiccup

Comment (johnkerl) [reply apis/r/R/SOMAExperimentAxisQuery.R:1061]: Also, passing along a pro-tip from @eddelbuettel you can have `df_name = c('obs', 'var')` -- and  then `df_name <- match.arg(df_name)` -- for added clarity to code-users as well as code-readers/code-maintainers

Comment (johnkerl) [reply apis/r/R/SOMAExperimentAxisQuery.R:1061]: likewise `columns` sounds like they are actual columns, leading to another cognitive hiccup when it turns out to be not the case: please `column_names`

File: apis/r/R/roxygen.R
Comment (johnkerl) [inline apis/r/R/roxygen.R:173]: Please spell out whatever rd is standing for

You do save yourself a few keystrokes typing it out here, as a one-time cost, but the added clarity to every reader forever after will be a net time savings

File: apis/r/R/roxygen.R
Comment (johnkerl) [inline apis/r/R/roxygen.R:192]: Similarly please spell out `md`

**Review by johnkerl — CHANGES_REQUESTED** [review]
Several changes requested, mainly around maintainability

After this I'll run this code on some various data files & may have more to say from that point of view later

Sorry again for the belated review

File: apis/r/R/roxygen.R
Comment (mojaveazure) [reply apis/r/R/roxygen.R:172]: It contains functions for use in documentation only; each function returns an Rd-formatted string and is designed to be called in `\Sexpr{}` in Roxygen comment blocks

File: apis/r/R/roxygen.R
Comment (mojaveazure) [reply apis/r/R/roxygen.R:173]: Rd stands for "R documentation", but is referred to as "Rd" in all manuals (such as ["Writing R Extensions"](https://cran.r-project.org/doc/manuals/R-exts.html#Rd-format) and ["Parsing Rd Files"](https://developer.r-project.org/parseRd.pdf)) and help pages (such as [`tools::parse_Rd()`](https://stat.ethz.ch/R-manual/R-devel/library/tools/html/parse_Rd.html)) within R itself

Comment (johnkerl) [reply apis/r/R/roxygen.R:173]: Then please add a top-of-function comment that this is an R-documentation generator.

This is quite obvious to the R-jedi (you being one) but this is a polyglottal project, with polyglottal developers as well as polyglottal users. Please just add an explanatory comment for the rest of us.

File: apis/r/R/roxygen.R
Comment (mojaveazure) [reply apis/r/R/roxygen.R:213]: `m` comes from `obsm` and `varm`; it could be `rd_outgest_m_layers`, but the `m` is a SOMA-specific suffix

**Review by johnkerl — COMMENTED** [review]
Some testing observations.

```
experiment <- SOMAExperimentOpen('/var/s/v/ileum')
experiment <- SOMAExperimentOpen('tiledb://johnkerl-tiledb/ileum')
```

both work well -- which is grand.

```
query <- SOMAExperimentAxisQuery$new(
  experiment = experiment,
  measurement_name = "RNA"
)

> length(query$obs_joinids())
[1] 32458
> length(query$var_joinids())
[1] 22068

obj <- query$to_sce()
```

results in

```
Error: Read of 'file:///var/s/v/ileum/ms/RNA/X/data' is incomplete.
Consider increasing the memory allocation via the configuration
option 'soma.init_buffer_bytes', or using iterated partial reads.
```

Trying this

```
platform_config <- PlatformConfig$new()
platform_config$set('tiledb', 'create', 'soma.init_buffer_bytes', 32*1024**3)
experiment <- SOMAExperimentOpen('tiledb://johnkerl-tiledb/ileum', platform_config=platform_config)
```

results in the same memory issue.

I thought to try a particular cell-type (using [this](https://github.com/single-cell-data/TileDB-SOMA/blob/1.2.4/apis/python/devtools/peek-exp.py#L22-L30)) with this output

```
cell_type
enteroendocrine cell of small intestine            18
paneth cell of epithelium of small intestine       34
transit amplifying cell of small intestine         53
smooth muscle fiber of ileum                       54
mast cell                                          92
endothelial cell of lymphatic vessel               96
pericyte cell                                     121
glial cell                                        175
ileal goblet cell                                 230
progenitor cell                                   382
endothelial cell                                  565
fibroblast                                        571
enterocyte of epithelium proper of ileum          809
innate lymphoid cell                             1382
mononuclear phagocyte                            1635
B cell                                           3183
plasma cell                                      3898
native cell                                      4203
alpha-beta T cell                               14957
```

and I found that `B cell` didn't do well even with a larger memory allocation:

```
query <- SOMAExperimentAxisQuery$new(
    experiment = experiment,
    measurement_name = "RNA",
    obs_query = SOMAAxisQuery$new(value_filter = 'cell_type == "B cell"')
)
obj <- query$to_sce()
```

```
Error: Read of 'file:///var/s/v/ileum/ms/RNA/X/data' is incomplete.
Consider increasing the memory allocation via the configuration
option 'soma.init_buffer_bytes', or using iterated partial reads.
```

 It worked fine for `mast cell` (which as we see above has 92 `obs` rows):

```
query <- SOMAExperimentAxisQuery$new(
    experiment = experiment,
    measurement_name = "RNA",
    obs_query = SOMAAxisQuery$new(value_filter = 'cell_type == "mast cell"')
)
obj <- query$to_sce()
```

```
> obj
class: SingleCellExperiment
dim: 22068 92
metadata(0):
assays(1): data
rownames(22068): var0 var1 ... var22066 var22067
rowData names(16): gene_ids mito ... feature_name feature_reference
colnames(92): obs275 obs531 ... obs32244 obs32339
colData names(33): obs_id n_counts ... ethnicity development_stage
reducedDimNames(1): UMAP_HM
mainExpName: RNA
altExpNames(0):
```

I think our inability to get the query result back even with 32GB buffer size, for such a small number of cells (B cell) is worrisome.

I'm not convinced that the 32GB buffer size is getting everywhere it needs to get.

I'd be OK with that getting analyzed on this PR, or, on a follow-up PR -- either way.

And I offer to debug it with you.

**Comment by johnkerl** [comment]
Concretely:

* Let's make sure (on this PR or a subsequent) the larger buffer size is being wired into everywhere it needs to go
* I'd love it if we could leverage the iterated-read on `to_single_cell_experiment` but I acknowledge SCE might not be able to _accept_ chunks of X

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (mojaveazure) [reply apis/r/R/SOMAExperimentAxisQuery.R:174]: This change makes use of shared documentation generators introduced in this PR

**Comment by johnkerl** [comment]
> I think our inability to get the query result back even with 32GB buffer size, for such a small number of cells (B cell) is worrisome.
> 
> I'm not convinced that the 32GB buffer size is getting everywhere it needs to get.
> 
> I'd be OK with that getting analyzed on this PR, or, on a follow-up PR -- either way.
> 
> And I offer to debug it with you.

I verified by `printf` in `libtiledbsoma/src/utils/column_buffer.cc` that indeed we are seeing the default 16MB buffer size even when I set

```

```
platform_config <- PlatformConfig$new()
platform_config$set('tiledb', 'create', 'soma.init_buffer_bytes', 32*1024**3)

experiment <- SOMAExperimentOpen('tiledb://johnkerl-tiledb/ileum', platform_config=platform_config)

query <- SOMAExperimentAxisQuery$new(
    experiment = experiment,
    measurement_name = "RNA",
    obs_query = SOMAAxisQuery$new(value_filter = 'cell_type == "B cell"')
)
obj <- query$to_sce()
```

I'll debug a bit more.

**Review by aaronwolen — CHANGES_REQUESTED** [review]
I haven't been able to fully review but wanted to post a couple of comments I had.

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:330]: please GC

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:345]: please GC

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [reply apis/r/R/SOMAExperimentAxisQuery.R:938]: @mojaveazure I maintain my ask

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [reply apis/r/R/SOMAExperimentAxisQuery.R:1018]: i retract this

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:1186]: please GC

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:1207]: please GC

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (johnkerl) [inline apis/r/R/SOMAExperimentAxisQuery.R:1247]: please GC

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (aaronwolen) [inline apis/r/R/SOMAExperimentAxisQuery.R:1160]: Looks like we're performing 2 separate reads here: once to retrieve ids from the designated index column and the once to read in the rest of the data. Can we consolidate to avoid the extra overhead?

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (aaronwolen) [inline apis/r/R/SOMAExperimentAxisQuery.R:1135]: A little more documentation here would benefit other devs because there some nuances here:

- column_names is polymorphic
- a data.frame with 0 columns is returned if column_names = FALSE
- all columns are retrieved if column_names = TRUE

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (aaronwolen) [inline apis/r/R/SOMAExperimentAxisQuery.R:1165]: Nit: I'm all for explicitly naming arguments when dealing with functions that have many or non-obvious parameters, but for something like `setdiff()` I personally find it diminishes readability.

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (aaronwolen) [inline apis/r/R/SOMAExperimentAxisQuery.R:1233]: Let's add a log statement prior to converting [CR]sparseMatrix to make it clear where the time is being spent.

File: apis/r/tests/testthat/test-SOMAExperiment-query-matrix-outgest.R
Comment (aaronwolen) [inline apis/r/tests/testthat/test-SOMAExperiment-query-matrix-outgest.R:169]: Why do we need both `$as_vector()` and `as.character()`?

File: apis/r/tests/testthat/test-SCEOutgest.R
Comment (aaronwolen) [reply apis/r/tests/testthat/test-SCEOutgest.R:1]: Totally agree! 

My only suggestion is that we wrap all of the code used to create these test experiments into a helper function since we're performing lagely the same steps in test-SeuratOutgest-object, test-SeuratOutgest-reduction, and now test-SCEOutgest.

Or perhaps we update the existing helper, `create_and_populate_experiment()`, to optionally populate the other slots.

File: apis/r/R/utils-single-cell-experiment.R
Comment (aaronwolen) [inline apis/r/R/utils-single-cell-experiment.R:1]: Since we'll likely add support for exporting to other bioc classes in the future I suggest we rename utils-single-cell-experiment.R to utils-bioc.R, as a common place for all bioconductor-related utilities.

**Review by aaronwolen — CHANGES_REQUESTED** [review]
I ran some tests on a few different datasets and this works great! It's very close but I made several comments that should be addressed before merging. 

Re the documentation generator functions, I'm personally on the fence. On one hand, I totally appreciate the desire to reduce content duplication across the package. On the other hand, I'm a little concerned the added complexity will make it difficult for others to contribute to the docs. It's a tricky balance to strike and I'm not sure what the right call is here but I'd be interested to what others think (@eddelbuettel, @johnkerl, @pablo-gar).

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (mojaveazure) [reply apis/r/R/SOMAExperimentAxisQuery.R:886]: There's a comment in the code highlighting the ommision of `varm_layers` from the parameter list (and has been since https://github.com/single-cell-data/TileDB-SOMA/pull/1279/commits/961aad3fca749b0a0b25f9d0b84beb240285a023)
image

File: apis/r/tests/testthat/test-SOMAExperiment-query-matrix-outgest.R
Comment (mojaveazure) [reply apis/r/tests/testthat/test-SOMAExperiment-query-matrix-outgest.R:169]: DelayedArray (dependency of SingleCellExperiment) defines `as.character.Array()`; this conflicts with casting of `arrow::Array` objects to character. Arrow only defines `as.character.ArrowDatum()` as its the parent class of `arrow::Array`; R's method resolution selects `DelayedArray:::as.character.Array()` over `arrow:::as.character.ArrowDatum()` as "Array" is a better match for "Array" than "ArrowDatum"

Comment (aaronwolen) [reply apis/r/tests/testthat/test-SOMAExperiment-query-matrix-outgest.R:169]: Nice catch!

File: apis/r/R/SOMAExperimentAxisQuery.R
Comment (mojaveazure) [reply apis/r/R/SOMAExperimentAxisQuery.R]: Rebased and reworked `.load_m_axis()` to make use of expanded `to_sparse_matrix()` and `warn_if_dense()`

**Comment by pablo-gar** [comment]
@mojaveazure I'd be happy to test in the [Census acceptance tests](https://github.com/chanzuckerberg/cellxgene-census/blob/main/api/r/cellxgene.census/tests/testthat/acceptance-tests.R), let me know when it's ready for that type of testing and I will go ahead and implement the tests on our end.

File: apis/r/R/utils-single-cell-experiment.R
Comment (mojaveazure) [reply apis/r/R/utils-single-cell-experiment.R]: Reworked both into an expanded `check_package()` (updated the existing function with a `version` argument)

File: apis/r/tests/testthat/test-SCEOutgest.R
Comment (mojaveazure) [reply apis/r/tests/testthat/test-SCEOutgest.R:1]: `create_and_populate_experiment()` has been expanded to allow creation of `obsm`, `varm`, `obsp`, and `varp` layers; the updated functionality has been used for both Seurat and SCE tests

**Comment by mojaveazure** [comment]
@pablo-gar Awesome, that'd be great! I feel that this is ready for acceptance testing as it's passing all our internal tests and larger datasets to help stress-test would be fantastic. As this is R, it has all the same memory constraints as the Seurat outgestion, but functionally it should be ready to go

**Review by johnkerl — APPROVED** [review]
Accepting (but do not merge until @aaronwolen also accepts)

The issue with

```
platform_config <- PlatformConfig$new()
platform_config$set('tiledb', 'create', 'soma.init_buffer_bytes', 32*1024**3)
```

as noted above at https://github.com/single-cell-data/TileDB-SOMA/pull/1279#pullrequestreview-1442645344 -- I've since found (as narrated on #1423) is not with this PR but ratther with the config-setter -- the right way is not set-after-create, but rather set-on-create -- something like:

```
ctx <- tiledbsoma::SOMATileDBContext$new(c(soma.init_buffer_bytes="2000000000"))
```

**Review by aaronwolen — APPROVED** [review]
One more [comment](https://github.com/single-cell-data/TileDB-SOMA/pull/1279#discussion_r1263924578) but otherwise I think this is ready to go!

**Comment by pablo-gar** [comment]
Two things: good news and potentially some areas to improve scalability.

## Good news

Stress tests for the outgestor in the context of [Census](https://github.com/chanzuckerberg/cellxgene-census) indicate that:

1. All of the expected queries work.
2. The running times for large queries is in the same order as magnitude as `to_seurat`: 0.8 - 1.5x  SCE/Seurat.

See [interim results for the acceptances tests](https://github.com/chanzuckerberg/cellxgene-census/tree/5ae392a0cb8a9b035ec2c8f56a11bac98d35d4bd/api/r/cellxgene.census/tests) and a summary in this [spreadsheet](https://docs.google.com/spreadsheets/d/1dlp6ExRLUsMz7U020VvXjMiNEhtIFZmXsFVIwWj_yzk/edit#gid=1030770423)


## Opportunity to improve -- loading `X` with with nnz > 2^31-1

For very large queries that have `X` matrices with nnz > 2^31-1 I believe it will be possible to create `SingleCellExperiment` objects (albeit with enough memory).

For these cases this seem to be occurring (tested on a 5M cell x 60K gene query): 

1. `X` matrix is successfully created in `dgTMatrix` format via `tables()$read()$concat()`... kudos to Matrix package.
2. However the matrix is attempted to be converted to a `dgCMatrix` sometime during the outgestion process.
3. This leads to the following error from Matrix package

```
Error: Error in `asMethod(object)`: unable to coerce from TsparseMatrix to [CR]sparseMatrixwhen length of 'i' slot exceeds 2^31-1
```

**SingleCellExperiment is capable of handling `dgTMatrix` for the expression matrices, thus I believe we should at the very least provide the user with the opportunity to stay in that format to allow for loading of these `X` matrices with nnz > 2^31-1** 


@aaronwolen @mojaveazure thoughts?

**Comment by aaronwolen** [comment]
@pablo-gar, this is an excellent find! @mojaveazure, given this info I think it makes sense to default to dgT/COO unless the user explicitly asks for something else—and that option is something we could potentially tackle in a separate PR.

**Pull request merged**
